text/2017-10-06-api-layers.md

```diff
@@ -0,0 +1,211 @@
+    local: Local,
+}
+```
+
+You can create a new collector by creating a new `Global` data, and add a handle to a collector by
+sharing its `Global` data.  Using a `Handle`, you can create a `Scope` and pass it to a given
+function:
+
+```rust
+impl Collector {
+    pub fn new() -> Self;
+    pub fn handle(&self) -> Handle;
+}
+
+impl Handle {
+    pub fn pin<F, R>(&self, f: F) 
+    where F: FnOnce(&Scope) -> R {
+        self.local.pin(&self.global, f)
+    }
+
+    ... // other methods
+}
+```
+
+Note that `Handle::pin()`, on contrary to `Local::pin()`, does not require a reference to a
+`Global`, because `Handle` already has a reference.
+
+See [the
+implementation](https://github.com/jeehoonkang/crossbeam-epoch/blob/handle/src/collector.rs) for
+more details.
+
+
+## The default garbage collector API
+
+There will be the default collector and per-thread handle to the default collector:
+
+```rust
+lazy_static! { pub GLOBAL: Global = Global::new(); }
+thread_local! { pub HANDLE: Handle = Handle::new(); }
+
+struct Handle(Local);
+
+impl Handle {
+    fn new() -> Self {
+        Self { 0: Local::new(&GLOBAL) }
+    }
+}
+
+impl Deref for Handle {
+    type Target = Local;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl Drop for Handle {
+    fn drop(&mut self) {
+        unsafe { self.unregister(&GLOBAL); }
+    }
+}
+
+pub fn pin<F, R>(f: F) 
+where F: FnOnce(&Scope) -> R {
+    HANDLE.with(|handle| { handle.pin(&GLOBAL, f) })
+}
+```
+
+See [the implementation](https://github.com/jeehoonkang/crossbeam-epoch/blob/handle/src/default.rs)
+for more details.
+
+
+
+# Alternatives
+
+`&'scope Scope` vs. `Scope<'scope>`: Currently we pass `&'scope Scope` as a witness that the current
+participant is pinned. The reference type may incur runtime overhead of nested indirection, but this
+author believes the overhead if bearable. You may avoid this overhead by using `Scope<'scope>`
+instead as the witness, but it is an unorthodox choice.
+
+`Arc<Global>` vs. `&Global`: The proposed `Collector` is implemented as `Arc<Global>`, which incurs
+a runtime overhead of reference counting.  This author believes the overhead is negligible, because
+handle creation is likely in the cold path.  Using `&Global` instead of `Arc<Global>` might
+eliminate the runtime cost, but it will significantly complicate the API.
+
+
+
+# Unresolved questions
+
+For `#[no_std]` environment, a long-term plan would be separating out what is dependent on `std`,
+namely the default collector API, and what is not, namely the `Collector` API and the internal
+API. The proposed `Collector` and internal API is almost `std`-free except for the fact that (1) the
+`Collector` API uses `std::sync::Arc`, and (2) they rely on the data structures in the standard
+library. This author believes we can easily make it fully independent from `std` by (1)
+re-implementing `Arc` in `#[no_std]`, and (2) using a `#[no_std]` data structures after [the
```

Comment:
I'm hoping that `Arc` will eventually be provided in no-std (libcore, libcollections, etc) with an `Alloc` parameter. That's probably a ways away though.